File: src/main/kotlin/com/minami/datatypes/JavaTest.java
package com.minami.datatypes;

import java.io.IOException;

public class JavaTest {

    public static void main(String[] args) {
        Bus bus = new Bus("blue", "Mercedez", 2011);
        System.out.println(bus);
        // because of @JvmField, the property is not achieved by get
        System.out.println(bus.model);
        // even here, the property don't use set
        bus.model = "Volvo";
        // color does't have set because it's val
        System.out.println(bus.getColor());
        bus.setYear(2010);
        System.out.println(bus);

        // because of @JvmStatic there is no need of companion
        // Bus.Companion.busCompanion();
        Bus.busCompanion();
        // System.out.println(Bus.Companion.isAuto());
        System.out.println(Bus.isAuto);
        //Constant doesn't need annotation
        System.out.println(Bus.constant);

        // because of @JvmStatic there is no need of INSTANCE with object
        SingletonObj.doSomething();
        // SingletonObj.INSTANCE.doSomething();

        // For Java knowing the method could throw an exception,
        // it's needed to annotate the method with @Throws
        try {
            StaticBus.doIO();
        } catch(IOException e){
            System.out.println("IOException");
        }

        bus.printMe("teste");

        StaticBus.defaultArgs("The number is: ", 40 );
        StaticBus.defaultArgs("The number is: ");
    }
}

File: src/main/kotlin/com/minami/datatypes/JavaTest.kt
@file:JvmName("StaticBus")

package com.minami.datatypes

import java.io.IOException

data class Bus(val color: String, @JvmField var model: String, var year: Int) {

    companion object {
        const val constant = 25
        @JvmField val isAuto = false
        @JvmStatic fun busCompanion () = println("I'm in Car's companion object")
    }
    fun printMe(argument: String) = println("argument is $argument")


}

object SingletonObj {
    @JvmStatic fun doSomething() = println("I'm doing something in the singleton object")
}

@Throws(IOException::class)
fun doIO() {
    throw IOException()
}

// function with optional parameter, because it considers 25 default value
// The JvmOverloads generates both functions with num and without num
@JvmOverloads
fun defaultArgs(str: String, num: Int = 25) {
    println("$str $num")

}